libs/utils/src/filter.c

```diff
@@ -673,32 +705,81 @@ static bool celix_filter_matchSubString(const celix_filter_t* filter, const celi
-    case CELIX_FILTER_OPERAND_LESS:
-        return celix_filter_compareAttributeValue(filter, entry) < 0;
-    default:
-        assert(filter->operand == CELIX_FILTER_OPERAND_LESSEQUAL);
-        return celix_filter_compareAttributeValue(filter, entry) <= 0;
+    } else if (filter->operand == CELIX_FILTER_OPERAND_APPROX) {
+        return celix_filter_matchApprox(filter, entry);
+    }
+
+    assert(filter->operand == CELIX_FILTER_OPERAND_EQUAL || filter->operand == CELIX_FILTER_OPERAND_GREATER ||
+           filter->operand == CELIX_FILTER_OPERAND_LESS || filter->operand == CELIX_FILTER_OPERAND_GREATEREQUAL ||
+           filter->operand == CELIX_FILTER_OPERAND_LESSEQUAL);
+
+
+    //match for array types
+    if (entry->valueType == CELIX_PROPERTIES_VALUE_TYPE_LONG_ARRAY && filter->internal->convertedToLong) {
+        return celix_utils_matchLongArrays(filter->operand, entry->typed.arrayValue, filter->internal->longValue);
+    } else if (entry->valueType == CELIX_PROPERTIES_VALUE_TYPE_DOUBLE_ARRAY && filter->internal->convertedToDouble) {
+        return celix_utils_matchDoubleArrays(filter->operand, entry->typed.arrayValue, filter->internal->doubleValue);
+    } else if (entry->valueType == CELIX_PROPERTIES_VALUE_TYPE_BOOL_ARRAY && filter->internal->convertedToBool) {
+        return celix_utils_matchBoolArrays(filter->operand, entry->typed.arrayValue, filter->internal->boolValue);
+    } else if (entry->valueType == CELIX_PROPERTIES_VALUE_TYPE_VERSION_ARRAY && filter->internal->convertedToVersion) {
+        return celix_utils_matchVersionArrays(filter->operand, entry->typed.arrayValue, filter->internal->versionValue);
+    } else if (entry->valueType == CELIX_PROPERTIES_VALUE_TYPE_STRING_ARRAY) {
+        return celix_utils_matchStringArrays(filter->operand, entry->typed.arrayValue, filter->value);
+    }
+
+    //regular compare -> match
+    int cmp;
+    if (entry->valueType == CELIX_PROPERTIES_VALUE_TYPE_LONG && filter->internal->convertedToLong) {
+        cmp = celix_filter_cmpLong(entry->typed.longValue, filter->internal->longValue);
+    } else if (entry->valueType == CELIX_PROPERTIES_VALUE_TYPE_DOUBLE && filter->internal->convertedToDouble) {
+        cmp = celix_filter_cmpDouble(entry->typed.doubleValue, filter->internal->doubleValue);
+    } else if (entry->valueType == CELIX_PROPERTIES_VALUE_TYPE_BOOL && filter->internal->convertedToBool) {
+        cmp = celix_filter_cmpBool(entry->typed.boolValue, filter->internal->boolValue);
+    } else if (entry->valueType == CELIX_PROPERTIES_VALUE_TYPE_VERSION && filter->internal->convertedToVersion) {
+        cmp = celix_version_compareTo(entry->typed.versionValue, filter->internal->versionValue);
+    } else {
+        // type string or property type and converted filter attribute value do not match ->
+        // fallback on string compare
+        cmp = strcmp(entry->value, filter->value);
     }
+    return celix_utils_convertCompareToBool(filter->operand, cmp);
```

Comment:
Consider the example given in Doxygen:

>  Example: The filter "(key>20)" and a property set with a long value 3 for key "key", will match and the same  filter but with a property set which has a string value "3" for key "key", will not match.

For a property set that has value "3" for key "key", it will match because as a string "3" > "20". On the other hand, A property set has value 3 for key "key" will NOT match since 3 < 20. Just contrary to what's said in documentation.